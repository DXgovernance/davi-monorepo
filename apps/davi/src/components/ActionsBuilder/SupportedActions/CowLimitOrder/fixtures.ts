import { DecodedCall, SupportedAction } from 'components/ActionsBuilder/types';
import { BigNumber, utils } from 'ethers';
import { MOCK_ADDRESS } from 'hooks/Guilds/ens/fixtures';
import CowSettlement from 'contracts/CowSettlement.json';

const CowSettlementContract = new utils.Interface(CowSettlement.abi);

export const limitOrderDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.COW_SWAP_LIMIT_ORDER,
  function: CowSettlementContract.getFunction('setPreSignature'),
  to: '0xfFb1cd0F95368DDd06D556161c5D3d9f0f4Fe6d2',
  value: BigNumber.from(0),
  args: {
    orderUid: '',
    signed: '',
  },
  optionalProps: {
    sellAmount: BigNumber.from('5000000000000000000'),
    buyToken: MOCK_ADDRESS,
    sellToken: MOCK_ADDRESS,
    functionSignature: '',
  },
};
