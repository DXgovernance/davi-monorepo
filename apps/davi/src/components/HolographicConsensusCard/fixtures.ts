import { BigNumber } from 'ethers';
import { IStakes } from './types';

export const mockProposalStakeDetails: IStakes = {
  for: [
    {
      id: '1',
      staker: '0xaf8eb8c3a5d9d900aa0b98e3df0bcc17d3c5f698',
      amount: BigNumber.from('100000000000000000000'),
    },
    {
      id: '2',
      staker: '0xc5b20ade9c9cd5e0cc087c62b26b815a4bc1881f',
      amount: BigNumber.from('100000000000000000000'),
    },
    {
      id: '3',
      staker: '0xc5b20ade9c9cd5e0cc087c62b26b815a4bc1881f',
      amount: BigNumber.from('336000000000000000000'),
    },
    {
      id: '4',
      staker: '0xc5b20ade9c9cd5e0cc087c62b26b815a4bc1881f',
      amount: BigNumber.from('236600000000000000000'),
    },
    {
      id: '5',
      staker: '0xc5b20ade9c9cd5e0cc087c62b26b815a4bc1881f',
      amount: BigNumber.from('47138000000000000000'),
    },
  ],
  against: [
    {
      id: '6',
      staker: '0x84eeb305da0a4309a696d43de9f79f04e66eb4f8',
      amount: BigNumber.from('100000000000000000000'),
    },
  ],
};

export const mockProposalStakeDetailsEmpty = {
  for: [],
  against: [],
};
