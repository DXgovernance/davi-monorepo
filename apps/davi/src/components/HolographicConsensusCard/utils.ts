import { BigNumber, FixedNumber } from 'ethers';
import { Proposal } from 'types/types.guilds.d';
import { IStakes } from './types';

export const calculateSpeedometerValue = (
  totalStaked: Proposal['totalStaked']
): number => {
  if (!totalStaked) return 0;

  const stakedAgainst = totalStaked[0];
  const stakedFor = totalStaked[1];

  const stakeDifference = stakedFor.sub(stakedAgainst);
  let percentage: number;

  if (stakeDifference.gt(0)) {
    percentage = stakeDifference.mul(100).div(stakedFor).toNumber();
  } else if (stakeDifference.eq(0)) {
    percentage = 0;
  } else {
    percentage = stakeDifference.mul(100).div(stakedAgainst).toNumber();
  }
  return 5000 + percentage * 50;
};

export const calculateBNPercentage = (
  value: BigNumber,
  percentage: string | number
) => {
  // percentage should be an integer (before being divided by 100)
  // if a percentage has decimals, it trims them
  // TODO: handle decimals?
  if (typeof percentage === 'string') {
    if (percentage.includes('.')) {
      const percentageSplit = percentage.split('.');
      percentage = percentageSplit[0];
    }
  }
  if (typeof percentage === 'number') {
    if (percentage % 1 !== 0) {
      percentage = Math.trunc(percentage);
    }
  }

  const valueFixed = FixedNumber.fromValue(value);
  const percentageFixed = FixedNumber.fromValue(BigNumber.from(percentage));
  const hundredFixed = FixedNumber.fromValue(BigNumber.from(100));

  const resultFixed = valueFixed
    .mulUnsafe(percentageFixed)
    .divUnsafe(hundredFixed);

  const resultFixedInteger = resultFixed.toString().split('.')[0];
  const resultBN = BigNumber.from(resultFixedInteger);
  return resultBN;
};

export const checkUserStakeOption = (
  userAddress: string,
  stakeDetails: IStakes
) => {
  const userStakeFor = stakeDetails.for.find(
    stake => stake.staker.toLowerCase() === userAddress.toLowerCase()
  );
  if (userStakeFor) return 'for';

  const userStakeAgainst = stakeDetails.against.find(
    stake => stake.staker.toLowerCase() === userAddress.toLowerCase()
  );
  if (userStakeAgainst) return 'against';

  return null;
};

export const calculatePotentialReward = (
  stakes: IStakes,
  userCurrentStake: BigNumber,
  userAddress: string,
  userOption: BigNumber,
  totalStaked: Proposal['totalStaked'],
  daoBounty: BigNumber
) => {
  if (userOption === undefined || userOption === null) return BigNumber.from(0);
  const userStakedOptionIndex: number = userOption.toNumber() - 1;
  let userStakedOptionString: string;
  if (userStakedOptionIndex === 0) userStakedOptionString = 'against';
  if (userStakedOptionIndex === 1) userStakedOptionString = 'for';
  if (!userStakedOptionString)
    throw new Error('Invalid userStakedOptionString');

  const stakedInUserVote: BigNumber =
    totalStaked[userStakedOptionIndex].add(userCurrentStake);
  const stakedInUserVoteFixed: FixedNumber =
    FixedNumber.fromValue(stakedInUserVote);

  if (stakedInUserVote.isZero()) return BigNumber.from(0);

  const totalStakedWithoutDaoBounty = userCurrentStake
    .add(totalStaked[0])
    .add(totalStaked[1])
    .sub(daoBounty);

  // Calculate what the user staked
  const previouslyStakedByUser = stakes[userStakedOptionString].reduce(
    (acc: BigNumber, cur) => {
      if (cur.staker.toLowerCase() === userAddress.toLowerCase()) {
        return acc.add(cur.amount);
      }
      return acc;
    },
    BigNumber.from(0)
  );

  const totalUserStake = previouslyStakedByUser.add(userCurrentStake);

  // Calculate the coefficient for the base reward

  const initialRewardCoefficient = FixedNumber.fromValue(
    totalStakedWithoutDaoBounty
  ).divUnsafe(stakedInUserVoteFixed);

  const initialReward = FixedNumber.fromValue(totalUserStake).mulUnsafe(
    initialRewardCoefficient
  );

  const initialRewardBN = BigNumber.from(
    initialReward.toString().split('.')[0]
  );

  // If the user voted for "NO", then the reward is just the initial reward
  if (userOption.eq(1)) return initialRewardBN;

  // If the user voted for "YES", then the reward is the initial reward + a percentage of the DAO bounty
  const daoBountyRewardCoefficient = FixedNumber.fromValue(daoBounty).divUnsafe(
    stakedInUserVoteFixed
  );

  const daoBountyReward = FixedNumber.fromValue(totalUserStake).mulUnsafe(
    daoBountyRewardCoefficient
  );

  const daoBountyRewardBN = BigNumber.from(
    daoBountyReward.toString().split('.')[0]
  );

  return initialRewardBN.add(daoBountyRewardBN);
};
