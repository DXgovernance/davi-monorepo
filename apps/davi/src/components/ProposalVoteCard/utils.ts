import { VoteOnProposalProps } from './types';

export const checkVotingPower = ({
  votingPowerAtProposalSnapshot,
  votingPowerAtProposalCurrentSnapshot,
}: VoteOnProposalProps) => {
  const hasNoVotingPower =
    votingPowerAtProposalSnapshot &&
    Number(votingPowerAtProposalSnapshot?.toString()) <= 0;

  const hasVotingPowerAtCurrentSnapshot =
    votingPowerAtProposalCurrentSnapshot &&
    Number(votingPowerAtProposalCurrentSnapshot?.toString()) > 0;
  return {
    hasNoVotingPower,
    hasVotingPowerAtCurrentSnapshot,
  };
};
