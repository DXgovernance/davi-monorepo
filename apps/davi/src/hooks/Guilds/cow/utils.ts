import { GNOSIS_ID, LOCALHOST_ID, MAINNET_ID } from 'utils';

export const isAvailableOnCow = (chainId: number) => {
  return [MAINNET_ID, GNOSIS_ID, LOCALHOST_ID].includes(chainId);
};
