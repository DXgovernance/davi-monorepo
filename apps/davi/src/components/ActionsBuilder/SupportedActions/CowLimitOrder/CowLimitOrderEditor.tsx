import { ActionEditorProps } from '..';
import { Button } from 'components/primitives/Button';
import { Controller, useForm } from 'react-hook-form';
import { Input } from 'components/primitives/Forms/Input';
import { TokenAmountInput } from 'components/primitives/Forms/TokenAmountInput';
import { BigNumber, FixedNumber, utils } from 'ethers';
import { useCallback, useEffect, useMemo, useState } from 'react';
import {
  Control,
  ControlLabel,
  ControlRow,
} from 'components/primitives/Forms/Control';
import { useTranslation } from 'react-i18next';
import { useNetwork } from 'wagmi';
import LimitOrder, {
  ValidateLimitOrderValues as LimitOrderValues,
} from './validateLimitOrder';
import { useTypedParams } from 'Modules/Guilds/Hooks/useTypedParams';
import { CowQuote, useCow } from 'hooks/Guilds/cow/useCow';
import {
  FieldError,
  Spacer,
  SwapQuoteError,
  UnitPriceContainer,
} from './CowLimitOrderEditor.styled';
import { BiRefresh } from 'react-icons/bi';
import { ERC20_APPROVE_SIGNATURE, getNetworkById } from 'utils';
import { DecodedCall, SupportedAction } from 'components/ActionsBuilder/types';
import {
  settlementContractAddress,
  vaultRelayerContractAddress,
} from 'hooks/Guilds/cow/config';
import ERC20 from 'contracts/ERC20.json';
import { Loading } from 'components/primitives/Loading';
import { useDebounce } from 'hooks/Guilds/useDebounce';
import { TokenPickerInput } from 'components/TokenPickerInput';
import { useERC20Info } from 'hooks/Guilds/erc20/useERC20Info';
import { Box } from 'components/primitives/Layout';
import { formatUnits } from 'ethers/lib/utils';
import useStringToBigNumber from 'hooks/Guilds/conversions/useStringToBigNumber';

const CowLimitOrderEditor: React.FC<ActionEditorProps> = ({
  decodedCall,
  onSubmit,
  isEdit,
}) => {
  const { t } = useTranslation();

  const { guildId } = useTypedParams();

  const { chain } = useNetwork();

  const nativeTokenSymbol = useMemo(() => {
    return getNetworkById(chain?.id).nativeAsset.symbol;
  }, [chain]);

  const { getQuote, createOrder, getNativePrice, error: cowError } = useCow();

  const [quote, setQuote] = useState<CowQuote>(null);
  const [isUnitPriceOverridden, setIsUnitPriceOverridden] = useState(false);

  const parsedData = useMemo(() => {
    if (!decodedCall) return null;

    /**
     * Only the sell amount will be used to calculate the quote since reversed quotes are not supported
     */
    return {
      buyToken: decodedCall.optionalProps.buyToken,
      sellToken: decodedCall.optionalProps.sellToken,
      sellAmount: decodedCall.optionalProps.sellAmount,
    };
  }, [decodedCall]);

  const { control, handleSubmit, getValues, watch } = useForm({
    resolver: LimitOrder,
    context: { t },
    defaultValues: parsedData,
  });

  const { buyToken, sellToken, sellAmount } = getValues();
  const watchSellAmount = watch('sellAmount');
  const watchSellToken = watch('sellToken');
  const watchBuyToken = watch('buyToken');

  const { data: sellTokenInfo } = useERC20Info(watchSellToken);
  const { data: buyTokenInfo } = useERC20Info(watchBuyToken);

  // useDebounce will make sure we're not spamming the cow api
  const debouncedSellAmount = useDebounce(watchSellAmount, 200);

  const [unitBuyPrice, setUnitBuyPrice] = useState<string>(null);

  const buyAmount = useMemo(() => {
    if (!quote || !unitBuyPrice || unitBuyPrice === '0') return '';

    try {
      return FixedNumber.from(1)
        .divUnsafe(FixedNumber.from(unitBuyPrice))
        .mulUnsafe(
          FixedNumber.from(
            formatUnits(quote?.sellAmount, sellTokenInfo?.decimals)
          )
        )
        .round(8)
        ?.toString();
    } catch (error) {
      console.log(error);
      return '0';
    }
  }, [quote, unitBuyPrice, sellTokenInfo]);

  const retrieveUnitprice = useCallback(async () => {
    return await getNativePrice(buyToken, buyTokenInfo?.decimals);
  }, [buyToken, getNativePrice, buyTokenInfo?.decimals]);

  const handleUnitPriceChange = (value: string) => {
    setUnitBuyPrice(value);
    setIsUnitPriceOverridden(true);
  };

  const isQuoteLoading = useMemo(() => {
    if (
      buyToken &&
      sellToken &&
      debouncedSellAmount &&
      !quote &&
      !isUnitPriceOverridden
    ) {
      return true;
    }

    return false;
  }, [quote, buyToken, debouncedSellAmount, sellToken, isUnitPriceOverridden]);

  useEffect(() => {
    if (
      !buyToken ||
      !sellToken ||
      !debouncedSellAmount ||
      quote ||
      isUnitPriceOverridden
    ) {
      return () => {};
    }

    let isActive = true;

    const requestQuote = async () => {
      try {
        const quote = await getQuote({
          buyToken: buyToken,
          sellToken: sellToken,
          sellAmount: BigNumber.from(debouncedSellAmount).toString(),
          receiver: guildId,
        });

        if (isActive) {
          setQuote(quote);
          // setIsUnitPriceOverridden(false);
        }
      } catch (e: any) {
        console.log('error getting quote ', e);
      }
    };

    requestQuote();

    return () => {
      isActive = false;
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    debouncedSellAmount,
    buyToken,
    sellToken,
    guildId,
    isUnitPriceOverridden,
  ]);

  useEffect(() => {
    if (!buyToken || isUnitPriceOverridden || !sellToken) {
      return () => {};
    }

    let isActive = true;

    const requestUnitPrice = async () => {
      try {
        // retrieve native price before getting the quote,
        // else if quote has an error will get it dismissed.
        const unitPrice = await retrieveUnitprice();

        if (isActive) {
          setUnitBuyPrice(unitPrice?.toString());
        }
      } catch (e: any) {
        console.log('error getting unit price ', e);
      }
    };

    requestUnitPrice();

    return () => {
      isActive = false;
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    buyToken,
    sellToken,
  ]);

  const buyAmountBN = useStringToBigNumber(buyAmount, buyTokenInfo?.decimals);

  const submitAction = async (values: LimitOrderValues) => {
    const orderId = await createOrder({
      ...quote,
      buyAmount: buyAmountBN?.toString(),
      sellAmount: sellAmount?.toString(),
    });
    const ERC20Contract = new utils.Interface(ERC20.abi);
    const cowApprovalCall: DecodedCall = {
      ...decodedCall,
      callType: SupportedAction.ERC20_APPROVE,
      to: values.sellToken,
      function: ERC20Contract.getFunction('approve'),
      args: {
        spender: vaultRelayerContractAddress,
        amount: values.sellAmount,
      },
      optionalProps: {
        functionSignature: ERC20_APPROVE_SIGNATURE,
      },
    };

    const limitOrderCall: DecodedCall = {
      ...decodedCall,
      to: settlementContractAddress,
      args: {
        orderUid: orderId,
        signed: true,
      },
      optionalProps: {
        sellAmount: values.sellAmount,
        buyToken: values.buyToken,
        sellToken: values.sellToken,
        functionSignature: '',
      },
    };

    if (isEdit) {
      // in case of edit mode we submit only one action that is being edited
      return onSubmit([limitOrderCall]);
    }

    return onSubmit([cowApprovalCall, limitOrderCall]);
  };

  return (
    <div>
      <form onSubmit={handleSubmit(submitAction, console.error)}>
        <ControlRow>
          <Controller
            name="sellToken"
            control={control}
            render={({ field: { ref, ...field }, fieldState }) => {
              const { error } = fieldState;
              return (
                <>
                  <Control>
                    <ControlLabel>
                      {t('actionBuilder.inputs.sellToken')}
                    </ControlLabel>

                    <TokenPickerInput
                      {...field}
                      placeholder={t('actionBuilder.inputs.sellToken')}
                      onChange={sellToken => {
                        setQuote(null);
                        setIsUnitPriceOverridden(false);
                        field.onChange(sellToken);
                      }}
                      value={field.value}
                      readOnly
                    />
                    {!!error?.message && (
                      <FieldError>{error.message}</FieldError>
                    )}
                  </Control>
                </>
              );
            }}
          />

          <Spacer />

          <Controller
            name="buyToken"
            control={control}
            render={({ field: { ref, ...field }, fieldState }) => {
              const { error } = fieldState;
              return (
                <>
                  <Control>
                    <ControlLabel>
                      {t('actionBuilder.inputs.buyToken')}
                    </ControlLabel>

                    <TokenPickerInput
                      {...field}
                      placeholder={t('actionBuilder.inputs.buyToken')}
                      onChange={buyToken => {
                        setQuote(null);
                        setIsUnitPriceOverridden(false);
                        field.onChange(buyToken);
                      }}
                      value={field.value}
                      readOnly
                    />
                    {!!error?.message && (
                      <FieldError>{error.message}</FieldError>
                    )}
                  </Control>
                </>
              );
            }}
          />
        </ControlRow>

        <ControlRow>
          <Controller
            name="sellAmount"
            control={control}
            render={({ field: { ref, ...field }, fieldState }) => {
              const { error } = fieldState;

              return (
                <Control>
                  <ControlLabel>
                    {t('actionBuilder.inputs.sellAmount')}
                  </ControlLabel>
                  <>
                    <ControlRow>
                      <TokenAmountInput
                        {...field}
                        decimals={sellTokenInfo?.decimals}
                        isInvalid={!!error?.message}
                        onChange={sellAmount => {
                          setIsUnitPriceOverridden(false);
                          setQuote(null);
                          field.onChange(sellAmount);
                        }}
                      />
                    </ControlRow>

                    {!!error?.message && (
                      <FieldError>{error.message}</FieldError>
                    )}
                  </>
                </Control>
              );
            }}
          />
        </ControlRow>

        <ControlRow>
          <Controller
            name="unitBuyPrice"
            control={control}
            render={({ field: { ref, ...field }, fieldState }) => {
              const { error } = fieldState;
              return (
                <>
                  <Control>
                    <UnitPriceContainer>
                      <ControlLabel>
                        {buyTokenInfo?.symbol
                          ? t('actionBuilder.inputs.pricePerToken', {
                              token: buyTokenInfo?.symbol,
                            })
                          : t('actionBuilder.inputs.unitBuyPrice')}{' '}
                        in{' '}
                        {isQuoteLoading ? (
                          <Loading
                            loading
                            text
                            skeletonProps={{ height: '1rem', width: '1.6rem' }}
                            style={{ marginLeft: '0.5rem' }}
                          />
                        ) : (
                          <>{nativeTokenSymbol}</>
                        )}
                      </ControlLabel>
                      {!isQuoteLoading && (
                        <UnitPriceContainer
                          onClick={async () => {
                            const unitPrice = await retrieveUnitprice();
                            setUnitBuyPrice(unitPrice?.toString());
                            setIsUnitPriceOverridden(false);
                          }}
                        >
                          <Box>Market Price</Box>
                          <BiRefresh size={20} />
                        </UnitPriceContainer>
                      )}
                    </UnitPriceContainer>
                    {isQuoteLoading ? (
                      <Loading
                        loading
                        text
                        skeletonProps={{ height: '2.6rem' }}
                      />
                    ) : (
                      <>
                        <ControlRow>
                          <Input
                            {...field}
                            placeholder={'0.0'}
                            value={unitBuyPrice}
                            onChange={e => {
                              handleUnitPriceChange(e.target.value);
                            }}
                          />
                        </ControlRow>
                        {!!error && <FieldError>{error.message}</FieldError>}
                      </>
                    )}
                  </Control>
                </>
              );
            }}
          />
        </ControlRow>

        <ControlRow>
          <>
            <Control>
              <ControlLabel>
                {t('actionBuilder.inputs.atLeastAmount')} {buyTokenInfo?.symbol}
              </ControlLabel>
              {isQuoteLoading ? (
                <Loading loading text skeletonProps={{ height: '2.6rem' }} />
              ) : (
                <>
                  <ControlRow>
                    <Input
                      placeholder={'0.0'}
                      value={buyAmount}
                      disabled={true}
                    />
                  </ControlRow>
                </>
              )}
            </Control>
          </>
        </ControlRow>

        {cowError && <SwapQuoteError>{cowError}</SwapQuoteError>}

        {isQuoteLoading ? (
          <Loading
            loading
            text
            skeletonProps={{ height: '2.6rem' }}
            style={{ margin: '1rem 0 0' }}
          />
        ) : (
          <Button
            m="1rem 0 0"
            fullWidth
            data-testid="submit-cowLimitOrder"
            type="submit"
            disabled={!quote}
          >
            {t('actionBuilder.action.saveAction')}
          </Button>
        )}
      </form>
    </div>
  );
};

export default CowLimitOrderEditor;
