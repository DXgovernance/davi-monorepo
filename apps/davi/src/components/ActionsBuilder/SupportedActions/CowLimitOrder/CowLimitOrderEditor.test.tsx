import { BigNumber } from 'ethers';
import { render } from 'utils/tests';
import CowLimitOrderEditor from './CowLimitOrderEditor';
import { MOCK_ADDRESS } from 'hooks/Guilds/ens/fixtures';
import { limitOrderDecodedCallMock } from './fixtures';

const mockBigNumber = BigNumber.from(100000000);

jest.mock('utils', () => ({
  getNetworkById: () => ({
    nativeAsset: {
      symbol: 'TST',
    },
  }),
  usePrevious: () => undefined,
}));

jest.mock('hooks/Guilds/ens/useENSAvatar', () => ({
  __esModule: true,
  default: () => ({
    imageUrl: 'wagmi',
  }),
}));

jest.mock('hooks/Guilds/erc20/useERC20Info', () => ({
  useERC20Info: () => ({
    name: 'Test ERC20',
    symbol: 'TEST',
    decimals: 18,
    totalSupply: mockBigNumber,
  }),
}));

jest.mock('hooks/Guilds/erc20/useAllERC20Balances', () => ({
  useAllERC20Balances: () => ({
    data: [],
  }),
}));

const mockChainId = 123456;
jest.mock('wagmi', () => ({
  useNetwork: () => ({ chain: { id: mockChainId } }),
  useAccount: () => ({ address: MOCK_ADDRESS }),
}));

jest.mock('hooks/Guilds/tokens/useTokenList', () => ({
  useTokenList: () => ({
    tokens: [],
  }),
}));

describe('CowLimitOrderEditor', () => {
  it('Should match snapshot', () => {
    const { container } = render(
      <CowLimitOrderEditor
        decodedCall={limitOrderDecodedCallMock}
        onSubmit={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
