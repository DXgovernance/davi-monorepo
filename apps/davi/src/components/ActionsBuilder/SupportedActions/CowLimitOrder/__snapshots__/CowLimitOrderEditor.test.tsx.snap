// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CowLimitOrderEditor Should match snapshot 1`] = `
.c6 {
  border: none;
  background: none;
  background-color: transparent;
  box-shadow: none;
  width: 100%;
  box-sizing: border-box;
  outline: none;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  width: 100%;
  margin-right: 0;
  border: 1px solid #303338;
  border-radius: 1.5rem;
  padding: 0.5rem 0.8rem;
  background-color: transparent;
  outline: 1px solid #737977;
}

.c4 input {
  color: #fff;
  font-family: 'Inter',-apple-system,BlinkMacSystemFont,'Segoe UI','Roboto','Oxygen','Ubuntu','Cantarell','Fira Sans','Droid Sans','Helvetica Neue',sans-serif;
  font-size: 14px;
  font-weight: 500;
  text-align: left;
}

.c4 input::-webkit-input-placeholder {
  color: #A1A6B0;
}

.c4 input::-moz-placeholder {
  color: #A1A6B0;
}

.c4 input:-ms-input-placeholder {
  color: #A1A6B0;
}

.c4 input::placeholder {
  color: #A1A6B0;
}

.c4:hover {
  outline: 1px solid #fff;
}

.c4:focus {
  outline: 2px solid #fff;
}

.c5 {
  margin-right: 0.3rem;
  color: #737977;
}

.c7 {
  margin-left: 0.3rem;
  color: #737977;
}

.c0 {
  box-sizing: 'border-box';
  min-width: 0;
  margin: 0;
  padding: 0;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 0.75rem 0;
  width: 100%;
}

.c3 {
  margin-bottom: 0.75rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  color: #A1A6B0;
  font-size: 14px;
  font-weight: 500;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c8 {
  margin-right: 1rem;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  cursor: pointer;
}

<div>
   
  <div>
    <form>
      <div
        class="c0 c1"
      >
        <div
          class="c0 c2"
        >
          <div
            class="c0 c3"
          >
            actionBuilder.inputs.sellToken
          </div>
          <div>
            <div
              class="c4"
            >
              <div
                class="c5"
              >
                <div>
                  <div
                    class="paper"
                    style="border-radius: 50px; display: inline-block; margin: 0px; overflow: hidden; padding: 0px; background-color: rgb(24, 151, 242); height: 18px; width: 18px;"
                  >
                    <svg
                      height="18"
                      width="18"
                      x="0"
                      xmlns="http://www.w3.org/2000/svg"
                      y="0"
                    >
                      <rect
                        fill="#2362E1"
                        height="18"
                        rx="0"
                        ry="0"
                        transform="translate(2.018998228945791 -3.0005497255688565) rotate(458.4 9 9)"
                        width="18"
                        x="0"
                        y="0"
                      />
                      <rect
                        fill="#F94301"
                        height="18"
                        rx="0"
                        ry="0"
                        transform="translate(-8.641945850428243 4.495697794961231) rotate(268.8 9 9)"
                        width="18"
                        x="0"
                        y="0"
                      />
                      <rect
                        fill="#FA7900"
                        height="18"
                        rx="0"
                        ry="0"
                        transform="translate(-5.10539291960705 16.582508932884398) rotate(117.3 9 9)"
                        width="18"
                        x="0"
                        y="0"
                      />
                      <rect
                        fill="#C8144A"
                        height="18"
                        rx="0"
                        ry="0"
                        transform="translate(-15.515940342410328 13.951529017583056) rotate(280.5 9 9)"
                        width="18"
                        x="0"
                        y="0"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <input
                class="c6"
                name="sellToken"
                placeholder="actionBuilder.inputs.sellToken"
                readonly=""
                value=""
              />
              <div
                class="c7"
              >
                <svg
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polyline
                    points="6 9 12 15 18 9"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          class="c0 c8"
        />
        <div
          class="c0 c2"
        >
          <div
            class="c0 c3"
          >
            actionBuilder.inputs.buyToken
          </div>
          <div>
            <div
              class="c4"
            >
              <div
                class="c5"
              >
                <div>
                  <div
                    class="paper"
                    style="border-radius: 50px; display: inline-block; margin: 0px; overflow: hidden; padding: 0px; background-color: rgb(24, 151, 242); height: 18px; width: 18px;"
                  >
                    <svg
                      height="18"
                      width="18"
                      x="0"
                      xmlns="http://www.w3.org/2000/svg"
                      y="0"
                    >
                      <rect
                        fill="#2362E1"
                        height="18"
                        rx="0"
                        ry="0"
                        transform="translate(2.018998228945791 -3.0005497255688565) rotate(458.4 9 9)"
                        width="18"
                        x="0"
                        y="0"
                      />
                      <rect
                        fill="#F94301"
                        height="18"
                        rx="0"
                        ry="0"
                        transform="translate(-8.641945850428243 4.495697794961231) rotate(268.8 9 9)"
                        width="18"
                        x="0"
                        y="0"
                      />
                      <rect
                        fill="#FA7900"
                        height="18"
                        rx="0"
                        ry="0"
                        transform="translate(-5.10539291960705 16.582508932884398) rotate(117.3 9 9)"
                        width="18"
                        x="0"
                        y="0"
                      />
                      <rect
                        fill="#C8144A"
                        height="18"
                        rx="0"
                        ry="0"
                        transform="translate(-15.515940342410328 13.951529017583056) rotate(280.5 9 9)"
                        width="18"
                        x="0"
                        y="0"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <input
                class="c6"
                name="buyToken"
                placeholder="actionBuilder.inputs.buyToken"
                readonly=""
                value=""
              />
              <div
                class="c7"
              >
                <svg
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polyline
                    points="6 9 12 15 18 9"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="c0 c1"
      >
        <div
          class="c0 c2"
        >
          <div
            class="c0 c3"
          >
            actionBuilder.inputs.sellAmount
          </div>
          <div
            class="c0 c1"
          >
            <div
              class="c4"
            >
              <div
                class="c5"
              />
              <input
                aria-label="numerical input"
                autocomplete="off"
                autocorrect="off"
                class="c6"
                inputmode="decimal"
                maxlength="79"
                minlength="1"
                name="sellAmount"
                pattern="^[0-9]*[.,]?[0-9]*$"
                placeholder="0.0"
                spellcheck="false"
                type="text"
                value="5.0"
              />
              <div
                class="c7"
              >
                <div
                  aria-label="clear number"
                  class="c0 c9"
                  data-testid="clear number"
                >
                  <svg
                    fill="none"
                    height="18"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="18"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <line
                      x1="18"
                      x2="6"
                      y1="6"
                      y2="18"
                    />
                    <line
                      x1="6"
                      x2="18"
                      y1="6"
                      y2="18"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="c0 c1"
      >
        <div
          class="c0 c2"
        >
          <div
            class="c0 c10"
          >
            <div
              class="c0 c3"
            >
              actionBuilder.inputs.unitBuyPrice
               
              in
               
              <div
                class=""
                style="margin-left: 0.5rem;"
              >
                <span
                  aria-busy="true"
                  aria-live="polite"
                >
                  <span
                    class="react-loading-skeleton"
                    style="width: 1.6rem; height: 1rem; --base-color: #333; --highlight-color: #555;"
                  >
                    ‌
                  </span>
                  <br />
                </span>
              </div>
            </div>
          </div>
          <div
            class=""
          >
            <span
              aria-busy="true"
              aria-live="polite"
            >
              <span
                class="react-loading-skeleton"
                style="height: 2.6rem; --base-color: #333; --highlight-color: #555;"
              >
                ‌
              </span>
              <br />
            </span>
          </div>
        </div>
      </div>
      <div
        class="c0 c1"
      >
        <div
          class="c0 c2"
        >
          <div
            class="c0 c3"
          >
            actionBuilder.inputs.atLeastAmount
             
          </div>
          <div
            class=""
          >
            <span
              aria-busy="true"
              aria-live="polite"
            >
              <span
                class="react-loading-skeleton"
                style="height: 2.6rem; --base-color: #333; --highlight-color: #555;"
              >
                ‌
              </span>
              <br />
            </span>
          </div>
        </div>
      </div>
      <div
        class=""
        style="margin: 1rem 0px 0px;"
      >
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="height: 2.6rem; --base-color: #333; --highlight-color: #555;"
          >
            ‌
          </span>
          <br />
        </span>
      </div>
    </form>
  </div>
   
</div>
`;
